### PR TITLE
drivers: led: max7219: add driver and sample

### PR DESCRIPTION
**Pull request by yanagiis**
This PR implements MAX7219 driver and provides sample.

### CODE REVIEW

File: drivers/led/Kconfig.max7219 [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2022 Jimmy Ou <yanagiis@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig MAX7219
+	bool "MAX7219 LED drivers"
+	depends on SPI
```
Comment (galak) [inline drivers/led/Kconfig.max7219:6]: Add a `default $(dt_compat_enabled,$(DT_COMPAT_MAXIM_MAX7219))`

(see `drivers/sensor/lm75/Kconfig` for an example)

File: tests/drivers/build_all/led/prj.conf [hunk 2]
```diff
@@ -2,3 +2,4 @@ CONFIG_TEST=y
 CONFIG_TEST_USERSPACE=y
 CONFIG_LED=y
 CONFIG_LED_GPIO=y
+CONFIG_MAX7219=y
```
Comment (galak) [inline tests/drivers/build_all/led/prj.conf:5]: This should be able to be removed when the `default` in Kconfig is added.

File: samples/drivers/led_max7219/prj.conf [hunk 3]
```diff
@@ -0,0 +1,4 @@
+CONFIG_LED=y
+CONFIG_LOG=y
+CONFIG_MAX7219=y
```
Comment (galak) [inline samples/drivers/led_max7219/prj.conf:3]: can be removed when `default` added to Kconfig.

File: drivers/led/Kconfig.max7219 [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+config MAX7219_INITIALIZATION_DELAY
+	int "MAX7219 initialization delay in milliseconds"
+	depends on MAX7219
+	default 0
+	range 0 2147483647
+	help
+	  This option delays MAX7219 initialization for the specified
+	  amount of milliseconds. This is used to allow MAX7219 get ready
+	  before initialing.
```
Comment (gmarull) [inline drivers/led/Kconfig.max7219:19]: why is this needed?

**Review by galak — CHANGES_REQUESTED** [review]
A few changes here and there.

File: drivers/led/max7219.c [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+static int max7219_led_init(const struct device *dev)
+{
+	const struct max7219_config *config = dev->config;
+	struct max7219_data *data = dev->data;
+	int ret;
+
+#if defined(CONFIG_MAX7219_INITIALIZATION_DELAY) && CONFIG_MAX7219_INITIALIZATION_DELAY > 0
+	k_msleep(CONFIG_MAX7219_INITIALIZATION_DELAY);
+#endif
```
Comment (gmarull) [inline drivers/led/max7219.c:182]: why? this will delay the whole system init

File: drivers/led/max7219.c [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+ * aaaaaaaa aaaaaaaa aaaaaaaa aabbbccc
+ *
+ * a: index of max7219
+ * b: index of digit
+ * c: index of segment
+ */
+
+#define MAX7219_INDEX(led) ((led) & (~BIT_MASK(6))) >> 6
+#define DIGIT_INDEX(led)   (((led)&BIT_MASK(6)) >> 3)
+#define SEGMENT_INDEX(led) ((led)&BIT_MASK(3))
```
Comment (gmarull) [inline drivers/led/max7219.c:63]: ` & `

File: dts/bindings/led/maxim,max7219.yaml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        - 11
+        - 12
+        - 13
+        - 14
+        - 15
+      description: Intensity for MAX7219.
+    num-cascading:
+      type: int
+      required: false
+      default: 1
```
Comment (gmarull) [inline dts/bindings/led/maxim,max7219.yaml:54]: ditto

File: samples/drivers/led_max7219/sample.yaml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+sample:
+  description: Demonstration of the MAX7219 LED driver
+  name: MAX7219 sample
+tests:
+  sample.drivers.led.max7219:
+    platform_allow: stm32_min_dev_black
```
Comment (gmarull) [inline samples/drivers/led_max7219/sample.yaml:6]: please filter based on `dt_compat_enabled`

File: samples/drivers/led_max7219/src/main.c [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/led.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/zephyr.h>
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(app);
```
Comment (gmarull) [inline samples/drivers/led_max7219/src/main.c:15]: please define app level log level. don't hardcode it.

File: samples/drivers/led_max7219/src/main.c [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+void main(void)
+{
+	const struct device *led_dev = DEVICE_DT_GET_ANY(maxim_max7219);
+	int i, ret;
+
+	if (led_dev == NULL) {
+		LOG_ERR("No device with compatible maxim,max7219 found");
+		return;
+	}
```
Comment (gmarull) [inline samples/drivers/led_max7219/src/main.c:29]: why not `DEVICE_DT_GET_ONE`?

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/led/Kconfig.max7219 (hunk 1 above)
Comment (yanagiis) [reply drivers/led/Kconfig.max7219:6]: OK. Thanks. I will add that.

File: dts/bindings/led/maxim,max7219.yaml [hunk 11]
```diff
@@ -0,0 +1,55 @@
+# Copyright (c) 2022 Jimmy Ou <yanagiis@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Maxim MAX7219 display driver
```
[the remark this answers is not in the document]
Comment (yanagiis) [reply dts/bindings/led/maxim,max7219.yaml:5]: OK. Thanks. My mistake. I will remove it.

File: drivers/led/max7219.c [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+{
+	const struct max7219_config *config = dev->config;
+	struct max7219_data *data = dev->data;
+	int ret;
+
+#if defined(CONFIG_MAX7219_INITIALIZATION_DELAY) && CONFIG_MAX7219_INITIALIZATION_DELAY > 0
+	k_msleep(CONFIG_MAX7219_INITIALIZATION_DELAY);
+#endif
+
+	if (!device_is_ready(config->spi.bus)) {
```
[the remark this answers is not in the document]
Comment (yanagiis) [reply drivers/led/max7219.c:184]: OK. Thanks. I will change it.

File: drivers/led/Kconfig.max7219 (hunk 4 above)
Comment (yanagiis) [reply drivers/led/Kconfig.max7219:19]: MAX7219 may not be ready when zephyr calls `max7219_led_init`.
Plus it only has MOSI pin, so I can't check its status.
I have no idea how to handle this situation in zephyr. 
Do you have any suggestions?

File: drivers/led/max7219.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (yanagiis) [reply drivers/led/max7219.c:184]: Thanks. I will change it.

File: drivers/led/max7219.c [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+
+#include <zephyr/drivers/led.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/zephyr.h>
+
+#define LOG_LEVEL CONFIG_LED_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(max7219);
```
[the remark this answers is not in the document]
Comment (yanagiis) [reply drivers/led/max7219.c:27]: OK, thanks. I will add it.

File: dts/bindings/led/maxim,max7219.yaml [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+compatible: "maxim,max7219"
+
+include: spi-device.yaml
+
+properties:
+    scan-limit:
+      type: int
+      required: false
+      default: 7
```
[the remark this answers is not in the document]
Comment (yanagiis) [reply dts/bindings/led/maxim,max7219.yaml:15]: OK. I think I should remove the default value. Thanks a lot.

File: dts/bindings/led/maxim,max7219.yaml [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+        - 6
+        - 7
+      description: |
+        Number of scanned digits for MAX7219.
+        Display digit 0 only when scan-limit is 0, digits 0 & 1 when
+        scan-limit is 1, and so on.
+    intensity:
+      type: int
+      required: false
+      default: 0
```
[the remark this answers is not in the document]
Comment (yanagiis) [reply dts/bindings/led/maxim,max7219.yaml:32]: OK. MAX7219 has a default value. I should remove the default value. Thanks a lot.

File: dts/bindings/led/maxim,max7219.yaml (hunk 7 above)
Comment (yanagiis) [reply dts/bindings/led/maxim,max7219.yaml:54]: OK. MAX7219 has a default value. I should remove the default value. Thanks a lot.

File: samples/drivers/led_max7219/sample.yaml (hunk 8 above)
Comment (yanagiis) [reply samples/drivers/led_max7219/sample.yaml:6]: I will add it. Thanks.

File: samples/drivers/led_max7219/src/main.c (hunk 9 above)
Comment (yanagiis) [reply samples/drivers/led_max7219/src/main.c:15]: OK. I will change it. Thanks.

File: samples/drivers/led_max7219/src/main.c (hunk 10 above)
Comment (yanagiis) [reply samples/drivers/led_max7219/src/main.c:29]: It's better than `DEVICE_DT_GET_ANY`. Thanks for your suggestion. I will change it.

**Comment by yanagiis** [comment]
Hi @simonguinot. Sorry, I don't get it.
This module is what I use.
[MAX7219 8-Digital Segment Digital LED Display](https://www.amazon.com/HiLetgo-MAX7219-8-Digital-Segment-Digital/dp/B01D0WSCJA/ref=sr_1_5?keywords=max7219&qid=1657126308&sr=8-5)
I thought it connects to LEDs?

**Comment by simonguinot** [comment]
> Hi @simonguinot. Thanks for your review.
> 
> Sorry, I don't get it. This module is similar to what I use. [MAX7219 8-Digital Segment Digital LED Display](https://www.amazon.com/HiLetgo-MAX7219-8-Digital-Segment-Digital/dp/B01D0WSCJA/ref=sr_1_5?keywords=max7219&qid=1657126308&sr=8-5) I thought it connects to LEDs?

Hi @yanagiis. I think I misled you by speaking of LCD. It is my mistake.

Let me clarify: the MAX72{19,21} is a display LED driver. It displays digits. And the LED API is not suited to display digit values.

But maybe you can use a display driver to display the digits ? See the drivers/display directory. It looks like there are LED matrix drivers among them.

I don't know at all the display API. So maybe I am all wrong and the MAX72{19,21} is not compatible. But please have to look.

File: samples/drivers/led_max7219/src/main.c [hunk 16]
```diff
@@ -0,0 +1,63 @@
+/*
```
Comment (nashif) [inline samples/drivers/led_max7219/src/main.c:1]: do we really need yet another copy of this sample for another driver? This seems to the same as led_lp3943. Either unify and make existing sample work for new driver or provide something that is new.

**Review by nashif — CHANGES_REQUESTED** [review]

**Comment by yanagiis** [comment]
Hi @simonguinot. 
I have taken look at the display driver and `nrf_led_matrix` driver and related [issue](https://github.com/zephyrproject-rtos/zephyr/issues/40208).

IMHO, the `Display API` is also not suited to display digit values when MAX7219 connects to 7-segment digits LED display.
But the `Display API` is more suitable than the led API when MAX7219 connects to the LED matrix.

I will rewrite it as a display driver.
Thanks again for your suggestion.

**Comment by simonguinot** [comment]
> Hi @simonguinot. I have taken look at the display driver and `nrf_led_matrix` driver and related [issue](https://github.com/zephyrproject-rtos/zephyr/issues/40208).
> 
> IMHO, the `Display API` is also not suited to display digit values when MAX7219 connects to 7-segment digits LED display. But the `Display API` is more suitable than the led API when MAX7219 connects to the LED matrix.
> 
> I will rewrite it as a display driver. Thanks again for your suggestion.

No problem !
